go-controller/hybrid-overlay/pkg/controller/node_linux.go

```diff
@@ -38,19 +43,28 @@ type NodeController struct {
 	// contains a map of pods to corresponding tunnels
 	tunMap      map[string]string
 	tunMapMutex sync.Mutex
+	// flow cache map of cookies to flows
+	flowCache map[string]*flowCacheEntry
+	flowMutex sync.Mutex
+	flowChan  chan bool
+	stopChan  <-chan struct{}
 }
 
 // NewNode returns a node handler that listens for node events
 // so that Add/Update/Delete events are appropriately handled.
 // It initializes the node it is currently running on. On Linux, this means:
 //  1. Setting up a VXLAN gateway and hooking to the OVN gateway
 //  2. Setting back annotations about its VTEP and gateway MAC address to its own object
-func NewNode(kube kube.Interface, nodeName string) (*NodeController, error) {
+func NewNode(kube kube.Interface, nodeName string, stopChan <-chan struct{}) (*NodeController, error) {
 	node := &NodeController{
 		kube:        kube,
 		nodeName:    nodeName,
 		tunMap:      make(map[string]string),
 		tunMapMutex: sync.Mutex{},
+		flowCache:   make(map[string]*flowCacheEntry),
+		flowMutex:   sync.Mutex{},
+		flowChan:    make(chan bool),
```

Comment:
so we can have multiple goroutines that can receive a pod update, which would each be trying to update flows and signal that a flow sync is required. We also have this flow sync routine waiting every 30 sec or for this signal. So to your point, yeah the pod additions will be blocked from sending the signal until flow sync finishes. Is that a bad thing? I'm kind of undecided.

The positive is that we can be fairly certain that when we see "pod wired up log message" that the next sync that happens is for that pod. So it might be easier to debug. 

The negative I see is let's say that we have 3 pod adds all taking place at the exact same time, and they all get to send this signal at the same time (not likely dude to the tunmap and flow locking, but for the sake of this case...). We would then have 3 flow syncs that happen, and 2 of these methods are blocked. However each flow sync would be for the same flows, since all 3 goroutines have finished updating their flows by the time they all send the signal, and we are blocking the methods as well as doing 2 additional syncs that are basically useless.

Maybe we should buffer it, and then somehow read off all of the values in the buffer at sync time, so that we only sync 1 time if multiple pod adds are finished? Is this overkill?

Reply:
the hack that `BoundedFrequencyRunner` uses is that it uses a channel of length 1, and then when pushing to it, it uses `select` to implement "push an element if there isn't already one there, but do nothing if there already is one" (https://github.com/kubernetes/kubernetes/blob/master/pkg/util/async/bounded_frequency_runner.go#L220). So then any number of "pushes" result in a single "pop".

Reply:
I see. I'll make it a buffered size of 1 then, and make the updating methods use a select like that. Thanks!